opal/corelib/kernel.rb

```diff
@@ -233,6 +248,72 @@ def extend(*mods)
     self
   end
 
+  def freeze
+    %x{
+      if (self.$$frozen) { return self; }
+      self.$$frozen = true;
+      try {
+        #{
+          if self.is_a?(String) || self.is_a?(Array)
+            return self
+          end
+        }
+        var type = typeof(self);
+        if (type === "object") {
+          // make setters of instance variables throw FrozenError
+          var prop, prop_type, current_value;
+          for(prop in self) {
+            prop_type = typeof(prop);
+            // prop_type "object" here is a String()
+            if ((prop_type === "string" || prop_type === "object") && prop.startsWith('$')) {
+              continue;
+            } else if (Number.isInteger(Number(prop))) {
+              // don't seal array indices
+              continue;
+            }
+            current_value = self[prop];
+            var ct = typeof(current_value);
+
+            // create closure for s and v
+            // for Opal 2.0 let for s and v should do the trick, instead of a function
+            (function() {
+              var s = false, p = prop, v;
+              Object.defineProperty(self, prop, {
+                get() { return v; },
+                set(value) {
+                  if (s) {
+                    console.error('---- dfa ----', p, self);
+                    $deny_frozen_access(self);
+                  }
+                  v = value;
+                  s = true;
+                }
+              });
+            })();
+            self[prop] = current_value; // must set value once
+          }
+
+          // set internal $$ properties for potential later access
+          if (!self.hasOwnProperty('$$comparable')) { $prop(self, '$$comparable', null); }
+          if (!self.hasOwnProperty('$$id')) { $prop(self, '$$id', null); }
+          if (!self.hasOwnProperty('$$meta')) { $prop(self, '$$meta', null); }
+          // These are for Regexp patterns
+          if (!self.hasOwnProperty('$$gm')) { $prop(self, '$$gm', null); }
+          if (!self.hasOwnProperty('$$g')) { $prop(self, '$$g', null); }
+
+          // seal the Object
+          Object.seal(self);
+        }
+      } catch(e) { console.err('freeze', e); }
```

Comment:
for debugging this is fine, but for final code this will need to be removed...

Reply:
Sure